drivers/serial/leuart_gecko.c

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright (c) 2018, Christian Taedcke, Diego Sueiro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <uart.h>
+#include <em_leuart.h>
+#include <em_gpio.h>
+#include <em_cmu.h>
+#include <board.h>
+
+#define DEV_CFG(dev) \
+	((const struct leuart_gecko_config * const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct leuart_gecko_data * const)(dev)->driver_data)
+#define DEV_BASE(dev) \
+	((LEUART_TypeDef *)(DEV_CFG(dev))->base)
+
+struct leuart_gecko_config {
+	LEUART_TypeDef *base;
+	CMU_Clock_TypeDef clock;
+	u32_t baud_rate;
+	struct soc_gpio_pin pin_rx;
+	struct soc_gpio_pin pin_tx;
+	unsigned int loc;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	void (*irq_config_func)(struct device *dev);
+#endif
+};
+
+struct leuart_gecko_data {
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t callback;
+	void *cb_data;
+#endif
+};
+
+static int leuart_gecko_poll_in(struct device *dev, unsigned char *c)
+{
+	LEUART_TypeDef *base = DEV_BASE(dev);
+	u32_t flags = LEUART_StatusGet(base);
+
+	if (flags & LEUART_STATUS_RXDATAV) {
+		*c = LEUART_Rx(base);
+		return 0;
+	}
+
+	return -1;
+}
+
+static unsigned char leuart_gecko_poll_out(struct device *dev, unsigned char c)
+{
+	LEUART_TypeDef *base = DEV_BASE(dev);
+
+	LEUART_Tx(base, c);
+
+	return c;
+}
+
+static int leuart_gecko_err_check(struct device *dev)
+{
+	LEUART_TypeDef *base = DEV_BASE(dev);
+	u32_t flags = LEUART_IntGet(base);
+	int err = 0;
+
+	if (flags & LEUART_IF_RXOF) {
+		err |= UART_ERROR_OVERRUN;
+	}
+
+	if (flags & LEUART_IF_PERR) {
+		err |= UART_ERROR_PARITY;
+	}
+
+	if (flags & LEUART_IF_FERR) {
+		err |= UART_ERROR_FRAMING;
+	}
+
+	LEUART_IntClear(base, LEUART_IF_RXOF |
+		       LEUART_IF_PERR |
+		       LEUART_IF_FERR);
+
+	return err;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static int leuart_gecko_fifo_fill(struct device *dev, const u8_t *tx_data,
+			       int len)
+{
+	LEUART_TypeDef *base = DEV_BASE(dev);
+	u8_t num_tx = 0;
+
+	while ((len - num_tx > 0) &&
+	       (base->STATUS & LEUART_STATUS_TXBL)) {
+
+		base->TXDATA = (u32_t)tx_data[num_tx++];
+	}
+
+	return num_tx;
+}
+
+static int leuart_gecko_fifo_read(struct device *dev, u8_t *rx_data,
+			       const int len)
+{
+	LEUART_TypeDef *base = DEV_BASE(dev);
+	u8_t num_rx = 0;
+
+	while ((len - num_rx > 0) &&
+	       (base->STATUS & LEUART_STATUS_RXDATAV)) {
+
+		rx_data[num_rx++] = (u8_t)base->RXDATA;
+	}
+
+	return num_rx;
+}
+
+static void leuart_gecko_irq_tx_enable(struct device *dev)
+{
+	LEUART_TypeDef *base = DEV_BASE(dev);
+	u32_t mask = LEUART_IEN_TXBL | LEUART_IEN_TXC;
```

Comment:
My understanding that this enables both "fifo has empty space" and "transmission complete" interrupts. Just a note that many other drivers don't enable "transmission complete" IRQ, because it's not really much useful.

Again, this is just a note of difference to other drivers. (If someone later would need to investigate any issues and path which led to them.
)

Reply:
I've just followed the same pattern for `drivers/serial/uart_gecko.c` driver.